proposals/lambda-improvements.md

```diff
@@ -72,18 +82,21 @@ It should be valid to inspect the `MethodInfo` associated with a `Delegate` inst
 This allows teams such as ASP.NET to make available the same behaviors for lambdas and local functions as ordinary methods.
 
 ## Explicit return type
-An explicit return type may be specified after the parameter list.
+An explicit return type may be specified before the parenthesized parameter list.
 ```csharp
-f = () : T => default;              // () : T
-f = x : short => 1;                 // <unknown> : short
-f = (ref int x) : ref int => ref x; // ref int : ref int
-f = static _ : void => { };         // <unknown> : void
+f = T () => default;                    // ok
+f = short x => 1;                       // syntax error at '=>'
+f = ref int (ref int x) => ref x;       // ok
```

Comment:
IIRC trailing return types were also chosen for function pointers because this is hard to read.

Couldn't we stick to `->` just like C++ with some restriction as to when you can specify the return type? e.g. only when parameters are also explicitly typed. See https://en.cppreference.com/w/cpp/language/lambda

Reply:
We discussed this and decided we really wanted to keep things consistent with all other function-forms in the language.  Methods are 'return type first'.  Local functions a 'return type first'.  So we would like to keep consistent with that approach.  

> e.g. only when parameters are also explicitly typed

This is a good question.  I don't think we covered this particular aspect @cston .  Should we say that you are only allowed to supply the return type if you also supply types for all parameters?  On hte one hand, i like the symettry of either having it be explicit or inferred.  On hte other hand though, i can see cases where users would want the param types inferred, but want to supply the explicit return type because it can't be effectively inferred.

Reply:
>Methods are 'return type first'. Local functions a 'return type first'.

I gave function pointers as an example where that's actually not true. Lambdas appear in the middle of an expression so if anything you'd expect the symmetry with the type it's being assigned to.

```cs
Func<A, B> M() { return (A p) -> B => e; }
Func<A, B> M() { return B(A p) => e; }
```
Even if I have to specify all the parameter types I'd def go with the former. That's just one way to disambiguate though.

Reply:
In my eyes both ways are reasonable - for short or built in types the "return type first syntax" looks quite good, too.
But on the other hand it looks quite similar to an invocation expression which decreases readability.

Reply:
> I gave function pointers as an example where that's actually not true.

Function pointers are a form of Type.  And it's true that in types we put return-last.  But lambdas are not really like types.  They're like an inline declaration of a method/local-function.  As such, we felt it should align most closely with that.  Same reasoning for making local functions look like methods.  

> But on the other hand it looks quite similar to an invocation expression which decreases readability.

Enh.  Only in the same way that a parenthesized lambda looks like a parenthesized expr.  You need to look forward to see if there are multiple params, or if you are followed by a `=>`.  Same as before.  So i think it's fine.  

Note that having the return type should be pretty rare.  only for cases where things break down and you really have no palatable alternative for the type to be understood.